### ISSUE TITLE
Monitor certificate expiry dates

### ISSUE THREAD

**Issue by emileswarts** [ministryofjustice/network-access-control-admin#456]
### User Story

As a network administrator
I need visibility over when certificates expire
So that I can plan ahead to rotate them 

### Value / Purpose

Improve visibility of certificate expiry times to reduce the chance of any service disruption.

### Useful Contacts

emile@madetech.com

### Additional Information

The expiry dates are already captured and stored in the admin as metadata.
The remaining months of each certificate needs to be turned into a metric.
This can be done by introducing a ECS background task that runs daily and prints this to cloudwatch logs.
This can then be pulled into Grafana, and panels display these remaining month metrics.
An alarm needs to be set up to send an alert into the network-access-control-critical-alerts slack channel when there are less than 4 months remaining.

In addition to this, we will display a banner in the admin when any of the certificates have less than 4 months remaining.
The remaining months needs to be persisted in a new table so heavy calculations aren't required on each page load.

### Definition of Done

Example - [ ] Documentation has been written / updated
- [ ] README has been updated
- [ ] User docs have been updated
- [ ] Another team member has reviewed
- [ ] Tests are green
- [ ] Grafana displays a new certificate expiry panel, with alarms set up
- [ ] A banner is displayed on the admin for any certificates with less than 4 months until they expire

**Comment by bagg3rs** [comment]
Waiting on @netops to decide on way to monitor certs.

**Comment by jennycasanova** [comment]
Meeting to be set up with Gateway Services to discuss ask before commencing development.  Meeting agreed with John Oparah.  Justin Fielding and Matti Sysmalainen wish to be included.  A representative will be invited from the NAC project team and Jenny Casanova will establish who at 01.11.22 meeting.  
Meeting needs to be as soon as possible due to time sensitivity of the ask from the project team and Gateway Services.

**Comment by jennycasanova** [comment]
Notification to be issued at 6 months in the NAC service alert slack channel and the Wifi and Lan slacks channel.
Notification to be issued at 3 months in the NAC critical alert channel in slack and the wifi and Lan team channel on slack.
Banner on the top of each page of Network Access Control Service pages on the admin portal.  For 6 months and 3 months.  Wording can be devs choice.
Email to be sent out at 6 months and 3 months in additional but the capability to change recipiant as needed due to length of time these notifications arise.

**Comment by jennycasanova** [comment]
Requirement discussion with PM - Shreya.  
Workflow behind the banner with status update and email notification all to be included in this ticket.  
- Warning banner appears when there is 3 months left on a certificate on the Network Access Control Service pages on the admin portal.
- Slack message issued in LAN team channel of the same warning when there is 3 months left on a certificate.  
- Status of banner updated to show that LAN engineer has actioned the request for certificate renew.  (if possible)
- Status of banner updated to done once certificate renew has been completed. (if possible)
- Email notification to engineers in additional to slack notification to ensure engineers not accessing the NACS portal go on and see the banner notification.

**Comment by jennycasanova** [comment]
Discussed with DevOp team.  Point 3 and 4 are not going to be possible to do as this will involve linking to servicenow which will be very time consuming to do for little benefit and value for a lot of dev work.
Clarify with Lan team if they want daily email and slack notifications rather than just one notification.

**Comment by jennycasanova** [comment]
Agreed with LAN team.  Ticket will be split.  This ticket will be to produce the banner onto the top of each page of the NAC portal when a certificate is due to expire at 4 months left.
Matti will confirm the wording for the banner which I will add to this ticket.

A second ticket will be created to cover the slack notification required.

**Comment by jennycasanova** [comment]
Devop team have improved the banner to say rather than a different banner for each certificate it will produce a banner when there is any or multiple certificates about to expire and a link on the banner to the relevant certificate.

**Comment by Gary-H9** [comment]
Work is currently captured in [this](https://github.com/ministryofjustice/network-access-control-admin/tree/add-invalid-cert-banner) branch.  

The private method which is called by the application view is called `certificate_expiry_check`, this needs further work, I've selected 3 months for testing but this can be easily altered as can the view itself. 

```
def certificate_expiry_check
    UpcomingExpiringCertificate.last.expiry_date <= Date.today + 3.months
end
```

Ultimately however the method can be called in application.html.erb:

```
<% if user_signed_in? and certificate_expiry_check %>
        <%= render "layouts/certificate_banner" %>
<% end %>
```

🌱 Currently I need to manually go onto the admin container and run `rails db:seed` to seed the table in the locally running database. Further reading needs to be done into the existing scripts and setup to understand how this works 

There is still some work to do on this piece:
- add testing 🤔
- the view can be altered, it's quite verbose 
- add a function to update the `UpcomingExpiringCertificates` table when a new certificate is loaded.
- investigate if there is a more efficient way of manipulating the `UpcomingExpiringCertificates` table. By this I mean, is`certificate_expiry_check` checking for the last entry in the database efficient? Or is it possible/desirable to only ever have one row in the table which is written over every time a new certificate which expires sooner than the existing entry in the database is added.

**Comment by Gary-H9** [comment]
I've reflected on the above method and decided adding a new table to the application database introduces more problems than it solves. 

A new `UpcomingExpiringCertificates` table would need to be aware of entries in the Certificates table. Initially I thought the new table would only need to be aware of the next certificate which is expiring but if this certificate is updated you need to handle this scenario _then_ update the table to be aware of the next expiring certificate from the Certificate table. 

Simultaneously you also need to handle new certificates being added to the system.

 This means you need to add any of:
- a background job would need to be added which would run every day and keep the table up to date
- new `after_save` condition in the model 
- other things I barely understand

All of this complication to avoid having the method look at the Certificate table directly each time a page loads seems burdensome so I've elected to simply look up Certificates. 

Testing this methodology locally with 10 certificates didn't result in any slowness when loading pages. 

The methodology as it stands now is: 

The `helper_method` - `certificate_expiry_check` exists in the `ApplicationController` 
```
def certificate_expiry_check
    Certificate.where('expiry_date >= ? AND expiry_date <= ?', Date.today, 3.months.from_now.to_date).exists?
end
```

It looks at every row in the Certificate table in the database and returns a true if it find at least one certificate with an `expiry_date` in the next three months.

It is called in every page as it is rendered in `application.html.erb` 

```
...
<% if user_signed_in? and certificate_expiry_check %>
        <%= render "layouts/certificate_banner" %>
<% end %>
...
```

Testing is captured in `view_certificate_spec.rb`: 

```
context "when the user is authenticated" do
   context "when a certificate exists but is out of date" do
      let!(:certificate) { create :certificate, expiry_date: Date.today + 2.months }

      it "shows a banner which notifies the user that a certificate is out of date" do
        visit root_path
        expect(page).to have_content "A certificate will expire in less that three months."
      end
   end
end
```

**Comment by Gary-H9** [comment]
This ticket is now complete. The branch has been merged into main and is confirmed to be working on Dev, which has a certificate which falls within the parameters set to display a banner.
